modules/runtime-common/src/main/scala/com/snowplowanalytics/snowplow/runtime/AppHealth.scala

```diff
@@ -0,0 +1,204 @@
+ *
+ * {{{
+ * _ <- appHealth.addRuntimeHealthReporter(EventsInputStream, sourceAndAckIsHealthy)
+ * }}}
+ *
+ * ==Application processing==
+ *
+ * Once the application enters its processing phase, you can set the health status for any runtime
+ * service or setup configuration
+ *
+ * {{{
+ * // After catching an exception in the bad events stream
+ * _ <- Logger[F].error(e)("Problem with bad rows output stream")
+ * _ <- appHealth.becomeUnhealthyForRuntimeService(BadRowsOutputStream)
+ *
+ * // After bad rows stream becomes healthy again
+ * _ <- Logger[F].debug("Bad rows output stream is ok")
+ * _ <- appHealth.becomeHealthyForRuntimeService(BadRowsOutputStream)
+ *
+ * // After catching an exception with the external setup configuration
+ * // Note this will send an alert webhook
+ * _ <- Logger[F].error(e)("Problem with the provided password")
+ * _ <- appHealth.becomeUnhealthyForSetup(BadPassword)
+ *
+ * // After successful connection to the externally configured services
+ * // Note this will send the first hearbeat webhook
+ * _ <- Logger[F].error(e)("Everything ok with the provided setup configuration")
+ * _ <- appHealth.becomeHealthyForSetup
+ * }}}
+ *
+ * The application processing code does not need to explicitly send any monitoring alert or adjust
+ * the health probe return code.
+ */
+class AppHealth[F[_]: Monad, SetupAlert, RuntimeService] private (
+  private[runtime] val setupHealth: SignallingRef[F, AppHealth.SetupStatus[SetupAlert]],
+  runtimeHealth: Ref[F, Map[RuntimeService, F[AppHealth.RuntimeServiceStatus]]]
+) extends AppHealth.Interface[F, SetupAlert, RuntimeService] {
+  import AppHealth._
+
+  def becomeHealthyForSetup: F[Unit] =
+    setupHealth.set(SetupStatus.Healthy)
+
+  def becomeUnhealthyForSetup(alert: SetupAlert): F[Unit] =
+    setupHealth.set(SetupStatus.Unhealthy(alert))
+
+  def becomeHealthyForRuntimeService(service: RuntimeService): F[Unit] =
+    runtimeHealth.update(_ - service)
+
+  def becomeUnhealthyForRuntimeService(service: RuntimeService): F[Unit] =
+    runtimeHealth.update(_ + (service -> Applicative[F].pure(RuntimeServiceStatus.Unhealthy)))
+
+  def addRuntimeHealthReporter(service: RuntimeService, reporter: F[RuntimeServiceStatus]): F[Unit] =
+    runtimeHealth.update(_ + (service -> reporter))
+
+  private[runtime] def unhealthyRuntimeServices: F[List[RuntimeService]] =
+    for {
+      asMap <- runtimeHealth.get
+      pairs <- asMap.toList.traverse { case (service, statusF) =>
+                 statusF.map((_, service))
+               }
+    } yield pairs.collect { case (RuntimeServiceStatus.Unhealthy, service) => service }
+}
+
+object AppHealth {
+
+  /**
+   * Subset of AppHealth interface required during app's event-processing phase
+   *
+   * This interface should be part of the App's Environment. Processing Specs can provide a custom
+   * implementation.
+   */
+  trait Interface[F[_], -SetupAlert, -RuntimeService] {
```

Comment:
Is this interface needed, because we don't want `addRuntimeHealthReporter` to be exposed to event processor in an app? 

If so, maybe `Map.empty[RuntimeService, F[AppHealth.RuntimeServiceStatus]` should be param of `init` function? Then we could pass there anything we want when creating environment. Would that change anything?

Reply:
It mainly exists so we can mock it in our Specs e.g. [here](https://github.com/snowplow-incubator/snowplow-bigquery-loader/blob/898ef7fa503371de299f314dc7a4534f06bb7531/modules/core/src/test/scala/com.snowplowanalytics.snowplow.bigquery/MockEnvironment.scala#L239-L249).

Regarding putting it in the `init` function.... I think I did it this way so I could initialize the health probe or initialize the event source in either order.  But even as I type this I realize it doesn't make any sense.  I'll switch it to put the reporters into `init`.